File: elixir/bob/bob.exs
defmodule Bob do
  def hey(text) do
    cond do
      Speech.is_question?(text) -> 
        "Sure."
      Speech.is_shouting?(text) -> 
        "Whoa, chill out!"
      Speech.is_silence?(text) ->
        "Fine. Be that way!"
      true ->
        "Whatever."
    end
  end
end

defmodule Speech do
  def is_silence?(text) do
    (text |> String.strip |> String.length) == 0
  end
  
  def is_shouting?(text) do
    !is_silence?(text) && (text == String.upcase text) && Regex.match?(~r/[^,?0-9\s]/, text)
  end
  
  def is_question?(text) do
    String.ends_with? text, "?"
  end  
end

File: ruby/bob/bob.rb
class Bob
  def self.hey(words)
    speech = Speech.new(words)
    bob    = Bob.new(speech)
    bob.reply
  end

  def initialize(speech)
    @speech = speech
  end

  attr_reader :speech

  def reply
    if speech.shouting?
      'Whoa, chill out!' 
    elsif speech.question?
      'Sure.'
    elsif speech.silence?
      'Fine. Be that way!'
    else
      'Whatever.'
    end
  end
end

class Speech < String
  def initialize(text)
    super text.to_s.strip
  end

  def shouting?
    !empty? && (self == self.upcase) && (/[^,?0-9\s]/.match(self) != nil)
  end

  def question?
    end_with?('?')
  end

  def silence?
    empty?
  end
end
